### PR TITLE
Add More information to EXPECT_NO_THROW/ASSERT_NO_THROW output

### PR DESCRIPTION
**Pull request by Zitrax**
Cherry picked back commit 0354ccb that went missing in 1b20bd1 due to merge
issues as reported in #2071.

Changed from original commit:

* Use const char* in the constructor instead of const std::string& to avoid
  compile issue (error: conversion from ‘const char [2]’ to non-scalar type
  ‘testing::internal::AdditionalMessage’ requested). This follows the style of
  the existing ConstCharPtr.

* Applied clang-format on the changes to avoid exceeding the line length limit.

(cherry picked from commit 0354ccb049ea2b9e4696adb8b5a013287daf0c33)

### CODE REVIEW

**Comment by Zitrax** [comment]
Since I just made minor modifications I let the original author stand from the cherry-pick.

@Jonny007-MKD does this seem fine with you?

**Comment by Zitrax** [comment]
@Jonny007-MKD Thanks.

I see you seem to have reviewed your original commit 0354ccb though, see the "files changed" tab here for my commit.

**Comment by Jonny007-MKD** [comment]
@Zitrax: I have compared both commits in 2 browser tabs and then commented my own one on the line lengths 🤦‍♂️

**Comment by Zitrax** [comment]
I see 👍

**Comment by gennadiycivil** [comment]
Thank your for this contribution. Could you please elaborate and provide more information on exactly the issue / problems this is trying to address.
What is the before/after effect this would create and how is this beneficial?

**Comment by Zitrax** [comment]
The problem is that when using EXPECT_NO_THROW or ASSERT_NO_THROW and the code throws the output from googletest is:

Expected: <statement> doesn't throw an exception.
Actual: it throws.

which is not very useful. The patch changes this to:

Expected: <statement> doesn't throw an exception.
Actual: it throws <message>

where message is taken from std::exception::what() which is very helpful to get some information on what was thrown.

Prior to this fix one often has to launch the debugger, create a coredump or add debug print messages to find out what was thrown while with it the message might be enough to pinpoint what happened without that debug session.

googletest should give us this information when it is available.

**Comment by gennadiycivil** [comment]
I see. So I changed the PR title to actually say what we are doing

File: googletest/include/gtest/internal/gtest-internal.h [hunk 1]
```diff
@@ -1338,17 +1353,21 @@ class FlatTuple
 
 #define GTEST_TEST_NO_THROW_(statement, fail) \
   GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
-  if (::testing::internal::AlwaysTrue()) { \
+  if (::testing::internal::AdditionalMessage message = ".") { \
```
Comment (gennadiycivil) [inline googletest/include/gtest/internal/gtest-internal.h]: I am missing why this change is beneficial? it appears that the change amounts to the same thing anyway. What is the advantage of adding oprator bool which evaluates to  return ::testing::internal::AlwaysTrue() and not comparing to the ::testing::internal::AlwaysTrue() directly as it was before?

File: googletest/include/gtest/internal/gtest-internal.h [hunk 2]
```diff
@@ -1338,17 +1353,21 @@ class FlatTuple
...
+    } \
     catch (...) { \
       goto GTEST_CONCAT_TOKEN_(gtest_label_testnothrow_, __LINE__); \
     } \
   } else \
     GTEST_CONCAT_TOKEN_(gtest_label_testnothrow_, __LINE__): \
-      fail("Expected: " #statement " doesn't throw an exception.\n" \
-           "  Actual: it throws.")
+      fail(("Expected: " #statement " doesn't throw an exception.\n" \
+           "  Actual: it throws" + message.get()).c_str())
```
Comment (gennadiycivil) [inline googletest/include/gtest/internal/gtest-internal.h]:
```suggestion
           "  Actual: it throws." + message.get()).c_str())
```

File: googletest/test/gtest_unittest.cc [hunk 3]
```diff
@@ -3803,6 +3804,10 @@ TEST(AssertionTest, ASSERT_NO_THROW) {
   EXPECT_FATAL_FAILURE(ASSERT_NO_THROW(ThrowAnInteger()),
                        "Expected: ThrowAnInteger() doesn't throw an exception."
                        "\n  Actual: it throws.");
+  EXPECT_FATAL_FAILURE(ASSERT_NO_THROW(ThrowAnException("blablubb")),
+                       "Expected: ThrowAnException(\"blablubb\") "
+                       "doesn't throw an exception.\n  "
+                       "Actual: it throws: blablubb");
```
Comment (gennadiycivil) [inline googletest/test/gtest_unittest.cc]:
```suggestion
                       "Actual: it throws.: blablubb");
```

File: googletest/include/gtest/internal/gtest-internal.h (hunk 1 above)
Comment (Jonny007-MKD) [reply googletest/include/gtest/internal/gtest-internal.h]: We need declare `message` in the scope of the `if`. It can't go into the body of the `if` because then we cannot access it in the `else`.

File: googletest/test/gtest_unittest.cc (hunk 3 above)
Comment (Jonny007-MKD) [reply googletest/test/gtest_unittest.cc]: In terms of punctuation this is wrong, isn't it? 😄 Why do you think it is better?

File: googletest/test/gtest_unittest.cc [hunk 4]
```diff
@@ -4540,6 +4545,10 @@ TEST(ExpectTest, EXPECT_NO_THROW) {
   EXPECT_NONFATAL_FAILURE(EXPECT_NO_THROW(ThrowAnInteger()),
                           "Expected: ThrowAnInteger() doesn't throw an "
                           "exception.\n  Actual: it throws.");
+  EXPECT_NONFATAL_FAILURE(EXPECT_NO_THROW(ThrowAnException("blablubb")),
+                          "Expected: ThrowAnException(\"blablubb\") "
+                          "doesn't throw an exception.\n  "
+                          "Actual: it throws: blablubb");
```
[the remark this answers is not in the document]
Comment (Jonny007-MKD) [reply googletest/test/gtest_unittest.cc]: That's reasonable :)

**Comment by Zitrax** [comment]
One travis build failed early before even compiling, some known issue?

File: googletest/include/gtest/internal/gtest-internal.h (hunk 1 above)
Comment (gennadiycivil) [inline googletest/include/gtest/internal/gtest-internal.h]: Lets add a comment to describe why this is here"
We need declare message in the scope of the if. It can't go into the body of the if because then we cannot access it in the else. It also can't go before the if block because than we have to make up some very unique name (line number is not enough)."

**Comment by gennadiycivil** [comment]
> One travis build failed early before even compiling, some known issue?

Travis does this some times

File: googletest/include/gtest/internal/gtest-internal.h [hunk 5]
```diff
@@ -1282,6 +1282,21 @@ class FlatTuple
   }
 };
 
+class AdditionalMessage {
```
Comment (gennadiycivil) [inline googletest/include/gtest/internal/gtest-internal.h]: Lets make this a struct with value being public, there is really no reason for access control.

**Comment by gennadiycivil** [comment]
@Zitrax  My sincere apologies for the delay. (Got very busy for a few days ) . We had an internal discussion, like the functionality. 
Could you please take a look at my comment, merging ConstCharPtr with new AdditionalMessage.

File: googletest/include/gtest/internal/gtest-internal.h (hunk 5 above)
Comment (Zitrax) [reply googletest/include/gtest/internal/gtest-internal.h]: I merged the functionality in AdditionalMessage into ConstCharPtr and used the same style for the new code. It could have been done the other way too I guess, not sure what makes most sense - but this way results in a smaller diff.

File: googletest/include/gtest/internal/gtest-internal.h [hunk 6]
```diff
@@ -1282,21 +1286,6 @@ class FlatTuple
...
 };
 
-class AdditionalMessage {
- public:
-  AdditionalMessage(const char* message) : value(message) {}
-  AdditionalMessage& operator=(const std::string& message) {
-    value = message;
-    return *this;
-  }
-  operator bool() const { return ::testing::internal::AlwaysTrue(); }
```
Comment (Zitrax) [inline googletest/include/gtest/internal/gtest-internal.h]: Just a note that the previous code here used AlwaysTrue(), I don't know what significance that has so I left the ConstCharPtr implementation as it was with a simple "return true;"

Comment (Jonny007-MKD) [reply googletest/include/gtest/internal/gtest-internal.h]: I don't know either, I just kept the AlwaysTrue() from the previous if statement

**Comment by Jonny007-MKD** [comment]
@gennadiycivil: you requested a change: ".". I suppose you didn't mean to do this :) How can this be resolved?

**Comment by gennadiycivil** [comment]
Please take a look at the fail:
https://travis-ci.org/google/googletest/jobs/500044517


> @gennadiycivil: you requested a change: ".". I suppose you didn't mean to do this :) How can this be resolved?

You suppose correctly, thank you, my bad. Please feel free to ignore that request for

**Comment by Zitrax** [comment]
> Please take a look at the fail:
> https://travis-ci.org/google/googletest/jobs/500044517

So it failed only osx-clang, while it passed in all other configurations. The output is:

```
[ RUN      ] ExpectTest.EXPECT_NO_THROW
/Users/travis/build/google/googletest/googletest/src/gtest.cc:707: Failure
Expected: 1 non-fatal failure containing "Expected: ThrowAnException("blablubb") doesn't throw an exception.
  Actual: it throws.: blablubb"
  Actual:
/Users/travis/build/google/googletest/googletest/test/gtest_unittest.cc:4548: Non-fatal failure:
Expected: ThrowAnException("blablubb") doesn't throw an exception.
  Actual: it throws.
```

I don't know why that would happen only there and I have no mac for further investigation.

**Comment by Jonny007-MKD** [comment]
Weird. Can wie re-run the build?

**Comment by Zitrax** [comment]
@gennadiycivil ping, can we trigger the build again to rule out some travis hickup?

**Comment by gennadiycivil** [comment]
> @gennadiycivil ping, can we trigger the build again to rule out some travis hickup?

Running, 
https://travis-ci.org/google/googletest/jobs/500044517

**Comment by Zitrax** [comment]
Seems consistent, it failed again.

Looking at the output as I understand it the problem is that for the EXPECT_NO_THROW test we except to see `"Actual: it throws.: blablubb"` but we get `"Actual: it throws."`. Looking at the code that could mean that `gtest_msg` was not set or was emptied.

Could this mean we have entered the `catch(...)` block instead of the `catch (const std::exception& e)` block?

But even if that happened shouldn't we have seen `"Actual: it throws.."`, since we initialize the value to "." ?

I also don't understand why EXPECT_NO_THROW fails and ASSERT_NO_THROW pass.

File: googletest/include/gtest/internal/gtest-internal.h [hunk 7]
```diff
@@ -1363,19 +1367,27 @@ constexpr bool InstantiateTypedTestCase_P_IsDeprecated() { return true; }
...
+// some very unique name (line number is not enough).
 #define GTEST_TEST_NO_THROW_(statement, fail) \
   GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
-  if (::testing::internal::AlwaysTrue()) { \
+  if (::testing::internal::ConstCharPtr gtest_msg = ".") { \
     try { \
       GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
     } \
+    catch (const std::exception& e) { \
+      gtest_msg = std::string(": ") + e.what(); \
```
Comment (Youw) [inline googletest/include/gtest/internal/gtest-internal.h]: no woulder CI is failing:
`gtest_msg` holds a pointer to a temporary object that is destroyed immediately after this statement - 
hello usage of dangling pointer :)

Original implementation with `AdditionalMessage` didn't have this issue, since it owned a copy of the message, which is fine.

My opinion - we need `AdditionalMessage` class back, maybe named more consistently with `ConstCharPtr`.
And `ConstCharPtr& ConstCharPtr::operator=(const std::string& message)` shouldn't be allowed, by definition of `class ConstCharPtr`.

File: googletest/include/gtest/internal/gtest-internal.h [hunk 8]
```diff
@@ -1363,19 +1367,27 @@ constexpr bool InstantiateTypedTestCase_P_IsDeprecated() { return true; }
     GTEST_CONCAT_TOKEN_(gtest_label_testthrow_, __LINE__): \
       fail(gtest_msg.value)
 
+// We need to declare the ConstCharPtr object in the scope of the if. It
+// can't go into the body of the if because then we cannot access it in the
+// else. It also can't go before the if block because than we have to make up
+// some very unique name (line number is not enough).
```
Comment (Youw) [inline googletest/include/gtest/internal/gtest-internal.h]: > It also can't go before the if block because than we have to make up some very unique name (line number is not enough).

This is not completely accurate - there is much deeper problem:
We cannot go before the if block because `ASSERT_NO_THROW(...)` should expand to a single statement, so it can be used as:
```cpp
if (some_condition)
    ASSORT_NO_THROW(some_expression) << "Additional error message";
```

**Comment by gennadiycivil** [comment]
Closing, looks inactive
